torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -139,93 +165,30 @@ static void print_flat_alloc(std::ostream& os, const Allocate* alloc) {
      << "[" << flat_size << "];" << std::endl;
 }
 
-void CudaPrinter::visit(const Free* v) {
-  Stmt* p = v->get_parent();
-  while (p) {
-    const For* for_v = dynamic_cast<const For*>(p);
-    if (for_v &&
-        (for_v->loop_options().is_gpu_block_index() ||
-         for_v->loop_options().is_gpu_thread_index())) {
-      return;
-    }
-    p = p->get_parent();
+void CudaPrinter::visit(const Allocate* v) {
+  if (cuda_analysis_->cross_block_bufs().count(v->buffer_var()) != 0) {
+    emitIndent();
+    os() << "__shared__ ";
+    print_flat_alloc(os(), v);
+    return;
   }
-  throw std::runtime_error("Global free not supported yet");
-}
 
-void CudaPrinter::visit(const Allocate* v) {
-  Stmt* p = v->get_parent();
-  while (p) {
-    const For* for_v = dynamic_cast<const For*>(p);
-    if (for_v) {
-      if (for_v->loop_options().is_gpu_block_index()) {
-        emitIndent();
-        os() << "__shared__ ";
-        print_flat_alloc(os(), v);
-        return;
-      } else if (for_v->loop_options().is_gpu_thread_index()) {
-        emitIndent();
-        print_flat_alloc(os(), v);
-        thread_local_bufs_.insert(v->buffer_var());
-        return;
-      }
-    }
-    p = p->get_parent();
+  if (cuda_analysis_->thread_local_bufs().count(v->buffer_var()) != 0) {
+    emitIndent();
```

Comment:
Minor: Maybe a TODO to handle the case where the size of the shared memory is dynamic. Then the dynamic size needs to be provided at the kernel launch time.